File: frontend/src/lib/documentClassifierBuilder.ts
/**
 * Build document-type classifier trees from user-friendly recognition signals.
 * Signal definitions are loaded from GET /api/rule-book/recognition-signals.
 */

import type {
  DocumentRuleCondition,
  DocumentRuleConditionGroup,
  DocumentTypeClassifier,
} from "@/lib/v5DocumentTypes";
import {
  pickGroups,
  signalConditionFromCatalog,
  supportingGuardConditions,
  weakSignalSet,
  type RecognitionSignalId,
} from "@/lib/recognitionSignalCatalog";

export type { RecognitionSignalId } from "@/lib/recognitionSignalCatalog";

export type ClassifierLayout = "any_signal" | "all_signals" | "supporting_doc" | "grouped";

const cond = (
  field: string,
  operator: DocumentRuleCondition["operator"],
  value: string
): DocumentRuleCondition => ({
  type: "condition",
  field,
  operator,
  value,
});

const orGroup = (children: DocumentRuleCondition[]): DocumentRuleConditionGroup => ({
  type: "group",
  operator: "OR",
  children,
});

const andGroup = (
  children: Array<DocumentRuleCondition | DocumentRuleConditionGroup>
): DocumentRuleConditionGroup => ({
  type: "group",
  operator: "AND",
  children,
});

/** Map each checkbox id to one classifier leaf condition. */
export function signalToCondition(signalId: RecognitionSignalId): DocumentRuleCondition {
  return (
    signalConditionFromCatalog(signalId) ?? cond("document_text", "contains", "")
  );
}

function buildGroupedIdentityTree(
  signalIds: RecognitionSignalId[]
): DocumentRuleConditionGroup | DocumentRuleCondition | null {
  const available = new Set(signalIds);
  const andChildren: Array<DocumentRuleCondition | DocumentRuleConditionGroup> = [];
  const used = new Set<RecognitionSignalId>();
  const weakSignals = weakSignalSet();
  const groups = pickGroups();

  for (const group of groups) {
    const matched = group.filter((id) => available.has(id));
    if (matched.length === 0) continue;
    matched.forEach((id) => used.add(id));
    if (matched.length === 1) {
      andChildren.push(signalToCondition(matched[0]));
    } else {
      andChildren.push(orGroup(matched.map(signalToCondition)));
    }
  }

  const ungrouped = signalIds.filter((id) => !used.has(id) && !weakSignals.has(id));
  if (ungrouped.length === 1) {
    andChildren.push(signalToCondition(ungrouped[0]));
  } else if (ungrouped.length > 1) {
    andChildren.push(orGroup(ungrouped.map(signalToCondition)));
  }

  for (const id of signalIds) {
    if (weakSignals.has(id)) {
      andChildren.push(signalToCondition(id));
    }
  }

  if (andChildren.length === 0) return null;
  if (andChildren.length === 1) return andChildren[0];
  return andGroup(andChildren);
}

export function buildClassifierRootGrouped(
  signalIds: RecognitionSignalId[]
): DocumentRuleConditionGroup {
  const tree = buildGroupedIdentityTree(signalIds);
  return tree?.type === "group"
    ? tree
    : tree
      ? andGroup([tree])
      : { type: "group", operator: "AND", children: [] };
}

export function buildClassifierRoot(
  signalIds: RecognitionSignalId[],
  layout: ClassifierLayout
): DocumentRuleConditionGroup {
  const leaves = signalIds.map(signalToCondition);
  if (leaves.length === 0) {
    return { type: "group", operator: "AND", children: [] };
  }

  if (layout === "grouped") {
    return buildClassifierRootGrouped(signalIds);
  }

  if (layout === "all_signals") {
    return andGroup(leaves);
  }

  if (layout === "supporting_doc") {
    const grouped = buildGroupedIdentityTree(signalIds);
    const identity =
      grouped?.type === "group"
        ? grouped
        : grouped
          ? orGroup([grouped])
          : orGroup(leaves);
    return andGroup([identity, ...supportingGuardConditions()]);
  }

  return orGroup(leaves);
}

export function buildClassifierFromSignals(
  signalIds: RecognitionSignalId[],
  layout: ClassifierLayout,
  options: { priority: number; confidence?: number; enabled?: boolean }
): DocumentTypeClassifier {
  return {
    enabled: options.enabled ?? true,
    priority: options.priority,
    confidence: options.confidence ?? 0.85,
    root: buildClassifierRoot(signalIds, layout),
  };
}

function conditionMatchesSignal(
  condition: DocumentRuleCondition,
  signalId: RecognitionSignalId
): boolean {
  const expected = signalToCondition(signalId);
  return (
    condition.field === expected.field &&
    condition.operator === expected.operator &&
    String(condition.value ?? "") === String(expected.value ?? "")
  );
}

function collectConditions(
  node: DocumentRuleCondition | DocumentRuleConditionGroup
): DocumentRuleCondition[] {
  if (node.type === "condition") return [node];
  return node.children.flatMap((child) => collectConditions(child));
}

/** Best-effort: recover simple-mode checkboxes from a stored classifier tree. */
export function parseSignalsFromClassifier(
  root: DocumentRuleConditionGroup,
  allowed: RecognitionSignalId[]
): RecognitionSignalId[] {
  const leaves = collectConditions(root);
  const found: RecognitionSignalId[] = [];
  for (const signalId of allowed) {
    if (leaves.some((leaf) => conditionMatchesSignal(leaf, signalId))) {
      found.push(signalId);
    }
  }
  return found;
}

export function isSupportingDocClassifier(root: DocumentRuleConditionGroup): boolean {
  const leaves = collectConditions(root);
  return (
    leaves.some((l) => l.field === "has_invoice_no" && l.value === "false") &&
    leaves.some((l) => l.field === "is_commercial_invoice" && l.value === "false")
  );
}

export function inferClassifierLayout(root: DocumentRuleConditionGroup): ClassifierLayout {
  if (root.operator === "AND" && isSupportingDocClassifier(root)) {
    return "supporting_doc";
  }
  if (root.operator === "AND" && root.children.every((c) => c.type === "condition")) {
    return "all_signals";
  }
  return "any_signal";
}

File: backend/app/services/classification/document_classifier_builder.py
"""Build document-type classifier trees from recognition signals (parity with frontend)."""

from __future__ import annotations

from typing import Any

from app.schemas.document_type import DocumentTypeClassifier
from app.services.classification.recognition_signal_registry import (
    SIGNAL_CONDITIONS,
    SIGNAL_PICK_GROUPS,
    SUPPORTING_GUARDS,
    WEAK_SIGNAL_IDS,
    signal_condition,
)

RecognitionSignalId = str
ClassifierLayout = str

WEAK_SIGNALS = WEAK_SIGNAL_IDS
_SIGNAL_CONDITIONS = SIGNAL_CONDITIONS


def _cond(signal_id: RecognitionSignalId) -> dict[str, Any]:
    spec = signal_condition(signal_id)
    if spec is None:
        return {"type": "condition", "field": "document_text", "operator": "contains", "value": ""}
    return {"type": "condition", **spec}


def _or_group(children: list[dict[str, Any]]) -> dict[str, Any]:
    return {"type": "group", "operator": "OR", "children": children}


def _and_group(children: list[dict[str, Any]]) -> dict[str, Any]:
    return {"type": "group", "operator": "AND", "children": children}


def build_classifier_root_grouped(signal_ids: list[RecognitionSignalId]) -> dict[str, Any]:
    available = set(signal_ids)
    and_children: list[dict[str, Any]] = []
    used: set[RecognitionSignalId] = set()

    for group in SIGNAL_PICK_GROUPS:
        matched = [signal_id for signal_id in group if signal_id in available]
        if not matched:
            continue
        used.update(matched)
        if len(matched) == 1:
            and_children.append(_cond(matched[0]))
        else:
            and_children.append(_or_group([_cond(signal_id) for signal_id in matched]))

    ungrouped = [signal_id for signal_id in signal_ids if signal_id not in used and signal_id not in WEAK_SIGNALS]
    if len(ungrouped) == 1:
        and_children.append(_cond(ungrouped[0]))
    elif len(ungrouped) > 1:
        and_children.append(_or_group([_cond(signal_id) for signal_id in ungrouped]))

    for signal_id in signal_ids:
        if signal_id in WEAK_SIGNALS:
            and_children.append(_cond(signal_id))

    if not and_children:
        return {"type": "group", "operator": "AND", "children": []}
    return _and_group(and_children)


def build_classifier_root(signal_ids: list[RecognitionSignalId], layout: ClassifierLayout) -> dict[str, Any]:
    if not signal_ids:
        return {"type": "group", "operator": "AND", "children": []}

    if layout == "grouped":
        return build_classifier_root_grouped(signal_ids)

    leaves = [_cond(signal_id) for signal_id in signal_ids]
    if layout == "all_signals":
        return _and_group(leaves)
    if layout == "supporting_doc":
        identity = build_classifier_root_grouped(signal_ids)
        if identity.get("type") == "condition":
            identity = _or_group([identity])
        return _and_group([identity, *SUPPORTING_GUARDS])
    return _or_group(leaves)


def build_classifier_from_signals(
    signal_ids: list[RecognitionSignalId],
    layout: ClassifierLayout,
    *,
    priority: int = 100,
    confidence: float = 0.85,
    enabled: bool = True,
) -> DocumentTypeClassifier:
    return DocumentTypeClassifier(
        enabled=enabled and bool(signal_ids),
        priority=priority,
        confidence=confidence,
        root=build_classifier_root(signal_ids, layout),
    )


def eval_recognition_signal(ctx: object, signal_id: RecognitionSignalId) -> bool:
    """True when a single recognition signal's classifier condition matches the sample."""
    from app.services.classification.document_type_rule_engine import _document_field
    from app.services.rule_book.rule_engine import _match_value

    spec = signal_condition(signal_id)
    if spec is None:
        return False
    haystack = _document_field(ctx, str(spec["field"]))
    return _match_value(
        haystack,
        str(spec["operator"]),
        str(spec["value"]),
    )


def filter_verified_recognition_signals(
    ctx: object,
    signals: frozenset[RecognitionSignalId] | set[RecognitionSignalId],
) -> frozenset[RecognitionSignalId]:
    """Drop detection-only false positives before building a sample classifier."""
    verified = {signal_id for signal_id in signals if eval_recognition_signal(ctx, signal_id)}
    return frozenset(verified)
